metron-platform/Performance-tuning-guide.md

```diff
@@ -0,0 +1,326 @@
+    - spout parallelism
+    - bolt parallelism
+    - # executors
+Metron
+    - bolt cache size - handles how many messages can be cached. This cache is used while waiting for all parts of the message to be rejoined.
+
+## Topologies
+
+### Parsers
+
+The parsers and PCAP use a builder utility, as opposed to enrichments and indexing, which use Flux.
+
+We set the number of partitions for our inbound Kafka topics to 48.
+
+```
+$ cat ~metron/.storm/storm-bro.config
+
+{
+    ...
+    "topology.max.spout.pending" : 2000
+    ...
+}
+```
+
+These are the spout recommended defaults from Storm and are currently the defaults provided in the Kafka spout itself.
+In fact, if you find the recommended defaults work fine for you, then this file might not be necessary at all.
+```
+$ cat ~/.storm/spout-bro.config
+{
+    ...
+    "spout.pollTimeoutMs" : 200,
+    "spout.maxUncommittedOffsets" : 10000000,
+    "spout.offsetCommitPeriodMs" : 30000
+}
+```
+
+We ran our bro parser topology with the following options
+
+```
+/usr/metron/0.4.0/bin/start_parser_topology.sh -k $BROKERLIST -z $ZOOKEEPER -s bro -ksp SASL_PLAINTEXT
+    -ot enrichments
+    -e ~metron/.storm/storm-bro.config \
+    -esc ~/.storm/spout-bro.config \
+    -sp 24 \
+    -snt 24 \
+    -nw 1 \
+    -pnt 24 \
+    -pp 24 \
+```
+
+From the usage docs, here are the options we've used. The full reference can be found here - https://github.com/apache/metron/blob/master/metron-platform/metron-parsers/README.md
+```
+-e,--extra_topology_options <JSON_FILE>        Extra options in the form
+                                               of a JSON file with a map
+                                               for content.
+-esc,--extra_kafka_spout_config <JSON_FILE>    Extra spout config options
+                                               in the form of a JSON file
+                                               with a map for content.
+                                               Possible keys are:
+                                               retryDelayMaxMs,retryDelay
+                                               Multiplier,retryInitialDel
+                                               ayMs,stateUpdateIntervalMs
+                                               ,bufferSizeBytes,fetchMaxW
+                                               ait,fetchSizeBytes,maxOffs
+                                               etBehind,metricsTimeBucket
+                                               SizeInSecs,socketTimeoutMs
+-sp,--spout_p <SPOUT_PARALLELISM_HINT>         Spout Parallelism Hint
+-snt,--spout_num_tasks <NUM_TASKS>             Spout Num Tasks
+-nw,--num_workers <NUM_WORKERS>                Number of Workers
+-pnt,--parser_num_tasks <NUM_TASKS>            Parser Num Tasks
+-pp,--parser_p <PARALLELISM_HINT>              Parser Parallelism Hint
+```
+
+### Enrichment
+
+Kafka - partitions setup
+    bro topic set to 48 partitions (referenced in the parser settings above)
```

Comment:
worth indicating why 48, or generalising to a function, or reference to number above